Design for new tools for ROS 2 package cross-compilation

zmichaels11 opened the pull request:
Old PR: https://github.com/ros2/design/pull/233

Commit:
Design for colcon plugin got ROS 2 package cross-compilation

Design doc for task ros2/ros2#693

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Fix mdl linter issues

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

nuclearsandwich committed:
Update articles/cross_compilation_build_tools_design.md

Co-Authored-By: juanrh <juan.rodriguez.hortala@gmail.com>

Commit:
Add references to more precedents for cross-compilation

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Explicitly state that QEMU is used to run `rosdep install`

To create the workspace dependent sysroot

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Fix some spelling and grammar mistakes

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Clarify the proposal just wraps existing cross compilaion techniques

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Replace colcon plugin by mixin and setup-sysroot command

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Extend the definition of platform beyond os and architecture

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Use at most one sentence per line

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Fix some typos and formatting issues

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Fix typos and use list to make the text easier to read

Signed-off-by: Juan Rodriguez Hortala <hortala@amazon.com>

Commit:
Address comments in PR #233

piraka9011 commented:
@dirk-thomas @pokitoz @nuclearsandwich @filiperinaldi Could you PTAL at this PR and recent updates?

I've updated the article based on your comments in #233. We've also opened a PR in the [`ros2/cross_compile`](https://github.com/ros2/cross_compile/pull/20) repo with the requested changes in design doc. 

Let us know if there's any updates to make. 
@prajakta-gokhale and I will address any issues moving forward.

dirk-thomas commented on articles/cross_compilation_build_tools_design.md:
```diff
@@ -0,0 +1,268 @@
+layout: default
+title: Cross-compiling ROS 2 packages
+abstract:
+  This article is a design proposal for extending the build tools for ROS 2 to
```
Just a nitpick about naming / phrasing:

In a over simplification the article describes:
* to use mixins to describe cross compilation configurations as well as
* a tool to setup / manage sysroot environments

Neither of these is something I would call "extend" the existing "build tool" (which is `colcon`). Therefore I think the abstract should be rephrased to give the reader a better idea what the article describes. The title of the PR (which will become the commit message when merged) should be updated accordingly. As a comparison the title of the article describes the scope better better imo.

pokitoz commented on articles/cross_compilation_build_tools_design.md:
```diff
@@ -0,0 +1,268 @@
+
+The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)
+and the [steve/ros2_raspbian_tools repository](https://github.com/esteve/ros2_raspbian_tools)
+contain instructions for cross-compiling ROS 2 to unsupported platforms like
```
nitpick about phrasing: arm64 and armhf are architectures and not platforms.

I would rephrase to:
>  [...] instructions for cross-compiling ROS 2 to unsupported architecture like arm64 and armhf

pokitoz commented on articles/cross_compilation_build_tools_design.md:
````diff
@@ -0,0 +1,268 @@
+  --packages-up-to performance_test # any other other `colcon build` arguments
+```
+
+- On an second iteration of this initiative we would:
````
> a

instead of

> an

pokitoz commented on articles/cross_compilation_build_tools_design.md:
```diff
@@ -0,0 +1,268 @@
+   from that base image and:
+   1. Runs `COPY` to get the contents of the workspace into the container.
+   1. Uses `rosdep` to ensure no workspace system dependency is missing.
+      Note we can run `rosdep` in Docker for platforms like ARM-HF thanks
```
I would update

> in Docker for platforms like ARM-HF

to 

> in Docker for architecture like ARM-HF

(architecture versus platform)

Commit:
Update for PR comments

- Removed mentions of old CMake toolchains.
- Updated abstract
- Updated features based on current implementation
- Addressed PR comments and grammar

piraka9011 commented:
@dirk-thomas @pokitoz I've updated the design doc with the requested changes along with some updates based on what we've done so far in the [ros2/cross_compile PR](https://github.com/ros2/cross_compile/pull/20).

dirk-thomas commented:
Looks good to me - but I don't have much insight into the content so others should "approve" this.

Nitpick: each sentence should start on a new line (following the style guide).

Commit:
Update styling

dirk-thomas commented on articles/cross_compilation_build_tools_design.md:
```diff
@@ -0,0 +1,170 @@
+abstract:
+  This article is a design proposal for developing a ROS 2 tool that sets up and
+  manages sysroot environments for cross-compilation with the objective of being simple
+  and extensible. Extending support for new cross compilation configurations using colcon 
```
Another case of one-sentence-per-line.

It also seems that several lines have trailing whitespaces which should be removed.

piraka9011 replied:
Are we using any linters for this? I went through the contributing guidelines and the linting checks all pass.

Updating in the next commit.

dirk-thomas replied:
Yes, they are supposed to catch this kind of stuff: https://github.com/ros2/design/blob/3f33a66aa5736c2557130181a2af72b0c37ca4f2/.travis.yml#L10

Commit:
Remove trailing whitespaces

piraka9011 commented:
If there aren't any other updates/suggestions could we get this merged/closed?

dirk-thomas commented:
Maybe someone should review / approve the content? Not sure who would be best for this. I was only looking at the content from a format / terminology point of view.

piraka9011 commented:
Pinging @pokitoz @alsora and @filiperinaldi to see if they have any suggestions.

drastx commented:
A lot of these arm and mips houses that make sbc development or reference boards use openwrt, buildroot or yocto. There are others, but those are the three major ones.

So a lot of the times, developers would have sysroot already built by openwrt or buildroot so being able to use that existing standard mechanism instead of creating specialized docker images would be helpful.

Doing it with docker, you're really hoping that the qemu image matches in terms of libc, special flags, and processor nuances for particular hardware but in reality, I think that will be difficult. The qemu arm support is a bit generalist in my opinion and can be used for platforms that are _well_ supported. But with any new platforms, you're basically limiting your portability because _you_ have to create a specialized docker image yourself. I doubt that any of the SBC houses will invest time doing that.

I used to maintain the tool-chain(compiler,linker,libc, etc...) for a major US-based arm/mips chip manufacturer and personally, I've never seen anyone using docker or qemu for cross-compilation. The approach seems really up the stack for very well supported platforms but not a right solution for hardware bring-up.

So, in my opinion, if you supported standard sysroot options like openwrt and buildroot etc then you are addressing the embedded market better.

In summary, it's ok if you want to use docker/qemu, but that shouldn't be the only supported sysroot. You should be able to point the build to say an openwrt or buildroot sysroot image. My personal feeling is that companies/projects that are doing new hardware support/enablement are more likely to use the openwrt or buildroot, or yocto. And then on the other hand docker approach is going to be used mostly by the audience who's just trying to figure out what this arm thing is... dipping their toes in for like a school project or alike.

tfoote commented:
It seems like we've resolved most of the initial issues. There are some additional suggestions/extension points suggested by @drastx above. 

I'd suggest that if no one objects we merge this as is in draft state and people can start working on it. And @drastx if you can make any specific suggestions to extend the proposal to add the other sysroots that would be valuable and can be reviewed on top of the current draft.

Commit:
Update permalink and tag as draft

thomas-moulard commented:
@tfoote We have updated the PR, let us know if it's OK and we'll merge it.

thomas-moulard commented:
@tfoote Merging as it meets the bar you set.
